cvprac/cvp_api.py

```diff
@@ -827,3 +833,60 @@ def remove_image_from_container(self, image, container):
                           'parentTask': ''}]}
         self._add_temp_action(data)
         return self._save_topology_v2([])
+
+    def deploy_device(self, device, container, configlets=None, image=None):
+        ''' Move a device from the undefined container to a target container.
+            Optionally apply device-specific configlets and an image.
+
+            Args:
+                device (dict): unique key for the device
+                container (dict): container info
+                configlets (list): list of dicts with configlet key/name pairs
+                image (dict): optional image info
+
+            Returns:
+                response (dict): A dict that contains a status and a list of
+                    task ids created (if any).
+
+                    Ex: {u'data': {u'status': u'success', u'taskIds': [u'32']}}
+        '''
+        info = 'Deploy device %s to container %s' % (device['fqdn'],
+                                                     container['name'])
+        self.log.debug(info)
+        # Add action for moving device to specified container
+        data = {'data': [{'id': 1,
+                          'info': info,
+                          'infoPreview': info,
+                          'action': 'update',
+                          'nodeType': 'netelement',
+                          'nodeId': device['key'],
+                          'toId': container['key'],
+                          'fromId': 'undefined_container',
+                          'nodeName': device['fqdn'],
+                          'toName': container['name'],
+                          'toIdType': 'container',
+                          'childTasks': [],
+                          'parentTask': ''}]}
+        try:
+            self._add_temp_action(data)
+        except CvpApiError as e:
+            if 'Data already exists' in e:
+                self.log.debug('Device %s already in container %s'
+                               % (device['fqdn'], container['name']))
+                pass
+
+        # Get proposed configlets device will inherit from container it is
+        # being moved to.
+        prop_conf = self.clnt.get('/provisioning/getTempConfigsByNetElementId.do?'
+                                  'netElementId=%s' % device['key'])
+        new_configlets = prop_conf['proposedConfiglets']
+        if configlets:
+            new_configlets.extend(configlets)
+        self.apply_configlets_to_device('deploy_device', device,
+                                        new_configlets, create_task=False)
+
+        # Apply image to the device
+        if image:
+            self.apply_image_to_device(image, device, create_task=False)
```

Comment:
Yes as of now a key value error will be raised when apply_image_to_device attempts to reference a key in a non dict object. I did this to be consistent with how the image is passed as its information dict to the other image functions. I'm not opposed to changing to handle the name instead of the dictionary.